core/src/main/java/io/grpc/CallOptions.java

```diff
@@ -197,47 +218,55 @@ public String getCompressor() {
   @Nullable
   @ExperimentalApi("https://github.com/grpc/grpc-java/issues/1767")
   public String getAuthority() {
-    return authority;
+    return getOption(AUTHORITY_KEY);
   }
 
   /**
    * Returns a new {@code CallOptions} with {@code executor} to be used instead of the default
    * executor specified with {@link ManagedChannelBuilder#executor}.
    */
   public CallOptions withExecutor(Executor executor) {
-    CallOptions newOptions = new CallOptions(this);
-    newOptions.executor = executor;
-    return newOptions;
+    return withOption(EXECUTOR_KEY, executor);
   }
 
   @Nullable
   public Executor getExecutor() {
-    return executor;
+    return getOption(EXECUTOR_KEY);
   }
 
+  /**
+   * Create a copy of this CallOptions object with key set to value.
+   */
+  public <T> CallOptions withOption(Key<T> key, T value) {
+    CallOptions newOptions = new CallOptions(this);
```

Comment:
If we don't continue using explicit fields for most options, then this doesn't work the same way since we won't be coping the `IdentityHashMap` in the constructor. I don't mind if we create a new copy of the map here, or we could keep the entries in a manually-created linked list and prepend for modification.